File: 3414-maximum-score-of-non-overlapping-intervals/3414-maximum-score-of-non-overlapping-intervals.c
#include <stdlib.h>
#include <string.h>

// Struct for sorting tracking
struct IntervalNode {
    int left;
    int right;
    int weight;
    int original_idx;
};

// State representation
struct State {
    long long score;
    int ids[4];
    int len;
};

// Sort comparator
int compareIntervals(const void* x, const void* y) {
    int r1 = ((struct IntervalNode*)x)->right;
    int r2 = ((struct IntervalNode*)y)->right;
    if (r1 < r2) return -1;
    if (r1 > r2) return 1;
    return 0;
}

// Lexicographical helper
int compareIds(int* a, int a_len, int* b, int b_len) {
    int min_len = a_len < b_len ? a_len : b_len;
    for (int i = 0; i < min_len; i++) {
        if (a[i] != b[i]) return a[i] - b[i];
    }
    return a_len - b_len;
}

void addId(int* src, int src_len, int val, int* dest) {
    int i = 0, j = 0;
    while (i < src_len && src[i] < val) {
        dest[j++] = src[i++];
    }
    dest[j++] = val;
    while (i < src_len) {
        dest[j++] = src[i++];
    }
}

struct State betterState(struct State a, struct State b) {
    if (a.score != b.score) {
        return a.score > b.score ? a : b;
    }
    return compareIds(a.ids, a.len, b.ids, b.len) <= 0 ? a : b;
}

int* maximumWeight(int** intervals, int intervalsSize, int* intervalsColSize, int* returnSize) {
    int n = intervalsSize;

    struct IntervalNode* a = (struct IntervalNode*)malloc(n * sizeof(struct IntervalNode));
    for (int i = 0; i < n; i++) {
        a[i].left = intervals[i][0];
        a[i].right = intervals[i][1];
        a[i].weight = intervals[i][2];
        a[i].original_idx = i;
    }

    qsort(a, n, sizeof(struct IntervalNode), compareIntervals);

    int* prev = (int*)malloc(n * sizeof(int));
    for (int i = 0; i < n; i++) {
        int lo = 0, hi = i - 1, pos = -1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (a[mid].right < a[i].left) {
                pos = mid;
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }
        prev[i] = pos;
    }

    // Multi-dimensional DP arrays allocation on heap to avoid stack limits
    struct State** dp = (struct State**)malloc(5 * sizeof(struct State*));
    for (int k = 0; k <= 4; k++) {
        dp[k] = (struct State*)calloc((n + 1), sizeof(struct State));
    }

    for (int k = 1; k <= 4; k++) {
        for (int i = 1; i <= n; i++) {
            int cur = i - 1;
            struct State skip = dp[k][i - 1];
            struct State base = dp[k - 1][prev[cur] + 1];

            struct State take;
            take.score = base.score + a[cur].weight;
            take.len = base.len + 1;
            addId(base.ids, base.len, a[cur].original_idx, take.ids);

            dp[k][i] = betterState(skip, take);
        }
    }

    *returnSize = dp[4][n].len;
    int* result = (int*)malloc((*returnSize) * sizeof(int));
    memcpy(result, dp[4][n].ids, (*returnSize) * sizeof(int));

    // Cleanup memory allocations
    free(a);
    free(prev);
    for (int k = 0; k <= 4; k++) free(dp[k]);
    free(dp);

    return result;
}

File: 3414-maximum-score-of-non-overlapping-intervals/3414-maximum-score-of-non-overlapping-intervals.js
/**
 * @param {number[][]} intervals
 * @return {number[]}
 */
var maximumWeight = function(intervals) {
    const n = intervals.length;

    // [left, right, weight, originalIndex]
    const a = new Array(n);
    for (let i = 0; i < n; i++) {
        a[i] = [intervals[i][0], intervals[i][1], intervals[i][2], i];
    }

    // Sort by right endpoint
    a.sort((x, y) => x[1] - y[1]);

    // prev[i] = last interval ending before a[i] starts
    const prev = new Array(n);
    for (let i = 0; i < n; i++) {
        let lo = 0;
        let hi = i - 1;
        let pos = -1;

        while (lo <= hi) {
            const mid = Math.floor(lo + (hi - lo) / 2);
            if (a[mid][1] < a[i][0]) {
                pos = mid;
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }
        prev[i] = pos;
    }

    // dp state management structure
    const createState = (score, ids) => ({ score, ids });

    const dp = new Array(5);
    for (let k = 0; k <= 4; k++) {
        dp[k] = new Array(n + 1);
    }

    for (let i = 0; i <= n; i++) dp[0][i] = createState(0, []);
    for (let k = 1; k <= 4; k++) dp[k][0] = createState(0, []);

    // Add value into a sorted array
    function add(ids, value) {
        const result = [];
        let i = 0;
        while (i < ids.length && ids[i] < value) {
            result.push(ids[i]);
            i++;
        }
        result.push(value);
        while (i < ids.length) {
            result.push(ids[i]);
            i++;
        }
        return result;
    }

    // Lexicographical array comparison
    function compare(arr1, arr2) {
        const minLen = Math.min(arr1.length, arr2.length);
        for (let i = 0; i < minLen; i++) {
            if (arr1[i] !== arr2[i]) return arr1[i] - arr2[i];
        }
        return arr1.length - arr2.length;
    }

    // Determine better state
    function better(stateA, stateB) {
        if (stateA.score !== stateB.score) {
            return stateA.score > stateB.score ? stateA : stateB;
        }
        return compare(stateA.ids, stateB.ids) <= 0 ? stateA : stateB;
    }

    for (let k = 1; k <= 4; k++) {
        for (let i = 1; i <= n; i++) {
            const cur = i - 1;
            const skip = dp[k][i - 1];
            const base = dp[k - 1][prev[cur] + 1];

            const ids = add(base.ids, a[cur][3]);
            const take = createState(base.score + a[cur][2], ids);

            dp[k][i] = better(skip, take);
        }
    }

    return dp[4][n].ids;
};
